File: inst/Matlab/fdaM/ftpsite/temp/derivs.m
function dy = derivs(tnow, y, wfd) 
% DERIVS sets up the 1st order system corresponding to   
%   linear differential operator defined by wfd.

% for example, if the system is of the second order,
%  y(1) is the current value of the variable
%  y(2) is the current value of the first derivative of the variable

%  last modified 31 October 2005

w  = eval_fd(tnow, wfd);
m  = length(w);
wmat = zeros(m, m);
wmat(1:(m-1),2:m) = eye(m-1);
wmat(m,:) = -w;
dy = wmat * y;

File: R/derivs.R
derivs <- function(tnow, y, wfd) {
	#  Sets up a linear differential equation of order m 
	#  as a system of m first order equations
	#  Arguments:
	#  TNOW ... A vector of values of the independent variable t
	#  Y    ... A matrix of m values of Y corresponding to TNOW
	#  WFD  ... A functional data object containing coefficient functions
	#  Returns:
	#  DY   ... A matrix of derivative values corresponding to Y
	
	#  Last modified:  24 March 2003
	
  w  <- eval.fd(tnow,wfd)
  m  <- length(w) - 1;
  wmat <- matrix(0, m, m)
  wmat[1:(m-1),2:m] <- diag(rep(1,m-1))
  wmat[m,] <- -w[2:(m+1)]
  dy <- wmat %*% y
  return(dy)
}
